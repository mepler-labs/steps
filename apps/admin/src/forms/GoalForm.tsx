import Button from 'atoms/Buttons/Button';
import Panel from 'atoms/Panel';
import Textarea from 'components/Form/Textarea';
import { Goal } from 'components/Goals/types';
import { Flex } from 'grid-styled';
import React from 'react';
import { Form } from 'react-final-form';

interface Props {
  onSubmit: (goal: Goal) => void;
  goal?: Goal;
}

const GoalForm: React.SFC<Props> = ({ onSubmit, goal }) => (
  <Form
    initialValues={{ goal: goal && goal.text }}
    onSubmit={onSubmit}
    render={({ handleSubmit }) => (
      <form onSubmit={handleSubmit}>
        <h2>{goal ? 'Edit Goal' : 'Add New Goal'}</h2>
        <Panel>
          <Textarea name="goal" placeholder="My goal is to..." label="goal" />
          <Flex justifyContent="center">
            <Button>Save</Button>
          </Flex>
        </Panel>
      </form>
    )}
  />
);

export default GoalForm;
