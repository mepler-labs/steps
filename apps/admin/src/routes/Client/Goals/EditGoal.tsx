import React from 'react';
import { connect } from 'react-redux';
import { History } from 'react-router';
import { bindActionCreators } from 'redux';

import { addAlert } from 'actions/alerts';
import { setClientGoals } from 'actions/clients';
import BackButton from 'atoms/Buttons/BackButton';
import Main from 'atoms/Main';
import { Alert, AlertLevel } from 'components/Alert/types';
import { Goal } from 'components/Goals/types';
import GoalForm from 'forms/GoalForm';
import { Client } from 'reducers/clients';

interface Props {
  actions: {
    addAlert(alert: Alert): void;
    setClientGoals(client: Client, goals: string[]): Promise<void>;
  };
  client: Client;
  history: History;
  goal: Goal;
}

export class EditGoal extends React.Component<Props> {
  updateGoal = goalData => {
    const updatedGoals = this.props.client.goals.map((goal, i) => {
      if (i == this.props.goal.id) {
        return goalData.goal;
      }
      return goal;
    });
    this.props.actions
      .setClientGoals(this.props.client, updatedGoals)
      .then(() => {
        this.props.history.push(`/clients/${this.props.client.id}/goals`);
        this.props.actions.addAlert({
          level: AlertLevel.Success,
          message: 'Success!',
          id: 'edit-goal-success',
        });
      })
      .catch(err => {
        this.props.actions.addAlert({
          level: AlertLevel.Error,
          message: err.message,
          id: 'new-goal-error',
        });
      });
  };

  render() {
    return (
      <Main>
        <BackButton to={`/goals`} />
        <GoalForm onSubmit={this.updateGoal} goal={this.props.goal} />
      </Main>
    );
  }
}

const mapStateToProps = ({ auth }, { match }) => ({
  client: auth.user,
  goal: {
    text: auth.user.goals[match.params.id],
    id: match.params.id,
  },
  // TODO: Use real goal ID instead of index
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ addAlert, setClientGoals }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditGoal);
