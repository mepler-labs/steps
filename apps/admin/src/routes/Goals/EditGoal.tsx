import React from 'react';
import GoalForm from 'forms/GoalForm';
import { Alert, AlertLevel } from 'components/Alert/types';
import BackButton from 'atoms/Buttons/BackButton';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { addAlert } from 'actions/alerts';
import { getClients, setClientGoals } from 'actions/clients';
import { Client } from 'reducers/clients';
import { History } from 'react-router';
import { withRouter } from 'react-router-dom';
import { Goal } from 'components/Goals/types';
import { findById } from 'helpers';
import Main from 'atoms/Main';

interface Props {
  actions: {
    addAlert(alert: Alert): void;
    getClients(): Promise<void>;
    setClientGoals(clientId: number, goals: string[]): Promise<void>;
  };
  client?: Client;
  history: History;
  goal: Goal;
}

export class EditGoal extends React.Component<Props> {
  componentWillMount() {
    this.props.actions.getClients().catch(err => {
      this.props.actions.addAlert({
        level: AlertLevel.Error,
        message: err.message,
        id: 'edit-goal-get-client-error',
      });
    });
  }

  updateGoal = goalData => {
    const updatedGoals = this.props.client.goals.map((goal, i) => {
      if (i == this.props.goal.id) {
        return goalData.goal;
      }
      return goal;
    });
    this.props.actions
      .setClientGoals(this.props.client.id, updatedGoals)
      .then(() => {
        this.props.history.push(`/clients/${this.props.client.id}/goals`);
        this.props.actions.addAlert({
          level: AlertLevel.Success,
          message: 'Success!',
          id: 'edit-goal-success',
        });
      })
      .catch(err => {
        this.props.actions.addAlert({
          level: AlertLevel.Error,
          message: err.message,
          id: 'new-goal-error',
        });
      });
  };

  render() {
    return (
      <Main>
        <BackButton to={`/clients/${this.props.client.id}/goals`} />
        <GoalForm onSubmit={this.updateGoal} goal={this.props.goal} />
      </Main>
    );
  }
}

const mapStateToProps = ({ clients }, { match }) => ({
  client: findById(clients.clients, match.params.id),
  goal: {
    text: findById(clients.clients, match.params.id).goals[match.params.goalId],
    id: match.params.goalId,
  },
  // TODO: Use real goal ID instead of index
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    { addAlert, getClients, setClientGoals },
    dispatch,
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(withRouter(EditGoal));
