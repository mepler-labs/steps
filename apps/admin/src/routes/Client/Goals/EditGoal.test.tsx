import { shallow } from 'enzyme';
import 'jest';
import React from 'react';
import GoalForm from 'forms/GoalForm';
import { EditGoal } from './EditGoal';

describe('EditGoal.tsx', () => {
  const client = {
    id: 123,
    goals: ['It is my goal to get a job'],
  };

  const history = {
    push: jest.fn(),
  };

  it('renders correctly', () => {
    const actions = {
      addAlert: jest.fn(),
      getClients: jest.fn().mockReturnValue(Promise.resolve()),
      setClientGoals: jest.fn().mockReturnValue(Promise.resolve()),
    };
    const wrapper = shallow(
      <EditGoal
        client={client}
        actions={actions}
        history={history}
        goal={{ id: 1, text: 'my goal' }}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('sets goals for a client', () => {
    const actions = {
      addAlert: jest.fn(),
      getClients: jest.fn().mockReturnValue(Promise.resolve()),
      setClientGoals: jest.fn().mockReturnValue(Promise.resolve()),
    };
    const wrapper = shallow(
      <EditGoal
        client={client}
        actions={actions}
        history={history}
        goal={{ id: 0, text: 'my goal' }}
      />,
    );

    const goalForm = wrapper.find(GoalForm);

    goalForm.simulate('submit', { goal: 'It is my goal to go to the moon' });

    expect(actions.setClientGoals).toHaveBeenCalledWith(client, [
      'It is my goal to go to the moon',
    ]);
  });

  it('displays an error if set goals fails', done => {
    const actions = {
      addAlert: jest.fn(),
      getClients: jest.fn().mockReturnValue(Promise.resolve()),
      setClientGoals: jest
        .fn()
        .mockReturnValue(Promise.reject({ message: 'some error' })),
    };
    const wrapper = shallow(
      <EditGoal
        client={client}
        actions={actions}
        history={history}
        goal={{ id: 1, text: 'my goal' }}
      />,
    );

    const goalForm = wrapper.find(GoalForm);

    goalForm.simulate('submit', {
      goal: 'It is my goal to go to the moon',
    });

    setTimeout(() => {
      expect(actions.addAlert).toHaveBeenCalled();
      done();
    }, 0);
  });

  it('redirects back to the goals page on success', async () => {
    const actions = {
      addAlert: jest.fn(),
      getClients: jest.fn().mockReturnValue(Promise.resolve()),
      setClientGoals: jest.fn().mockReturnValue(Promise.resolve()),
    };
    const wrapper = shallow(
      <EditGoal
        client={client}
        actions={actions}
        history={history}
        goal={{ id: 1, text: 'my goal' }}
      />,
    );

    const goalForm = wrapper.find(GoalForm);

    await goalForm.simulate('submit', {
      goal: 'It is my goal to go to the moon',
    });

    expect(history.push).toHaveBeenCalledWith('/clients/123/goals');
  });
});
