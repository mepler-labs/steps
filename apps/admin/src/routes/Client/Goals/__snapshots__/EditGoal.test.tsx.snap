// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditGoal.tsx renders correctly 1`] = `
<Main>
  <Styled(BackButton)
    to="/goals"
  />
  <GoalForm
    goal={
      Object {
        "id": 1,
        "text": "my goal",
      }
    }
    onSubmit={[Function]}
  />
</Main>
`;
